xds/internal/xdsclient/client_refcounted.go

```diff
@@ -40,19 +40,23 @@ var (
 
 func clientRefCountedClose(name string) {
 	clientsMu.Lock()
-	defer clientsMu.Unlock()
-
 	client, ok := clients[name]
 	if !ok {
 		logger.Errorf("Attempt to close a non-existent xDS client with name %s", name)
+		clientsMu.Unlock()
 		return
 	}
 	if client.decrRef() != 0 {
+		clientsMu.Unlock()
 		return
 	}
+	delete(clients, name)
+	clientsMu.Unlock()
+
+	// This attempts to close the transport to the management server and could
+	// take a while to complete. So, call it without holding the lock.
```

Comment:
I would rewrite this a little.  The reason to do it without the lock is not because it will "take a while" (because that implies it's purely for performance), but because the things it does could theoretically call into the `xdsclient` package again and deadlock.